### PR TITLE
fix(common): Cache Intl instances

### PR DESCRIPTION
**Pull request by Eddman**
## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[x] Other... Please describe: Performance fix
```

## What is the current behavior?

Every time date/number pipe is called a new instance of Intl.DateTime/NumberFormat is created. The constructors are however very expensive and in bigger screens it slows down the rendering.

## What is the new behavior?
Instances are stored into cache and reused. The rendering time went in my case from 23s to less than 100ms.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
Here for reference the exact same issue: https://github.com/yahoo/react-intl/issues/27

### CODE REVIEW

**Comment by ocombe** [comment]
Hello, the i18n pipes have been extensively changed in v5, we don't use the intl API anymore. Can you check if you still have the issue with v5, and if that's the case could you port your PR to the new pipes please? Thanks

**Comment by ocombe** [comment]
Closing this because of the lack of activity.

File: packages/common/src/pipes/intl.ts [hunk 1]
```diff
@@ -33,7 +35,16 @@ export class NumberFormatter {
       options.currency = typeof currency == 'string' ? currency : undefined;
       options.currencyDisplay = currencyAsSymbol ? 'symbol' : 'code';
     }
-    return new Intl.NumberFormat(locale, options).format(num);
+
+    // For performance cache Intl.NumberFormat instances. The constructor is very expensive
+    let key = locale + JSON.stringify(options);
+    let formatter = NUMBER_FORMAT_CACHE.get(key);
```
Comment (rossgardt) [inline packages/common/src/pipes/intl.ts]:
```js
const key = ...
const formatter = ...
```

File: packages/common/src/pipes/intl.ts [hunk 2]
```diff
@@ -134,8 +145,18 @@ function hourExtractor(inner: DateFormatterFn): DateFormatterFn {
   return function(date: Date, locale: string): string { return inner(date, locale).split(' ')[0]; };
 }
 
+const DATE_FORMAT_CACHE: Map<string, Intl.DateTimeFormat> = new Map();
+
 function intlDateFormat(date: Date, locale: string, options: Intl.DateTimeFormatOptions): string {
-  return new Intl.DateTimeFormat(locale, options).format(date).replace(/[\u200e\u200f]/g, '');
+  // For performance cache Intl.DateTimeFormat instances. The constructor is very expensive
+  let key = locale + JSON.stringify(options);
+  let formatter = DATE_FORMAT_CACHE.get(key);
```
Comment (rossgardt) [inline packages/common/src/pipes/intl.ts]:
```js
const key = ...
const formatter = ...
```

File: packages/common/src/pipes/intl.ts [hunk 3]
```diff
@@ -134,8 +145,18 @@ function hourExtractor(inner: DateFormatterFn): DateFormatterFn {
   return function(date: Date, locale: string): string { return inner(date, locale).split(' ')[0]; };
 }
 
+const DATE_FORMAT_CACHE: Map<string, Intl.DateTimeFormat> = new Map();
+
 function intlDateFormat(date: Date, locale: string, options: Intl.DateTimeFormatOptions): string {
-  return new Intl.DateTimeFormat(locale, options).format(date).replace(/[\u200e\u200f]/g, '');
+  // For performance cache Intl.DateTimeFormat instances. The constructor is very expensive
```
Comment (rossgardt) [inline packages/common/src/pipes/intl.ts]: IMO, we can omit this comment.
The naming of `DATE_FORMAT_CACHE` is explicit enough.

**Review by benedikt-roth — COMMENTED** [review]
Just started following the repo. Please keep this PR. I like it!

Some minor remarks, though.

Best,
Benji

**Comment by rossgardt** [comment]
> Hello, the i18n pipes have been extensively changed in v5, we don't use the intl API anymore. Can you check if you still have the issue with v5, and if that's the case could you port your PR to the new pipes please?

Sorry. Missed this message. Then my remarks might probably be outdated 🙂

**Comment by ocombe** [comment]
most definitely :)

**Pull request closed without merging**